File: tools/adversarial/translate.mjs
import {readFileSync} from 'node:fs';
import {compile} from '../../js/src/sel.mjs';
import {Binding, Sql} from '../../js/src/sql/index.mjs';
const bindings = Object.fromEntries(['r','s'].map(t=>[t.toUpperCase(),Binding.relation(t,t,Object.fromEntries(['id','cat','v','fk'].map(c=>[c.toUpperCase(),Binding.column(c,t,c==='cat'?'TEXT':'NUM')])))]));
const replay=process.env.AUDIT_REPLAY?readFileSync('tools/adversarial/replay-js.sel','utf8').trimEnd().split('\n'):null;
let n=0;
for(const [i,source] of readFileSync('tools/adversarial/queries.sel','utf8').trimEnd().split('\n').entries()) {
 for(const d of ['sqlite','postgresql','mariadb']) for(const strict of [false,true]) {
  const out={i,d,strict}; const p=compile(source);
  try {const f=Sql.translateStatement(p,d,bindings,{strict}); Object.assign(out,{sql:f.asStatement(),params_sql:f.asStatement('params'),params:f.bindings().map(v=>v.asText()),caveats:f.caveats});} catch(e){out.error=e.code??String(e);}
  try {const p1=Sql.planHybrid(p,d,bindings,{strict});out.plan=p1.pureSql?'pure_sql':p1.pureMemory?'pure_memory':'hybrid';out.prefix=p1.sqlStatement?.asStatement()??null;} catch(e){out.plan_error=e.code??String(e);}
  if(replay){const data=compile(replay[n]).run();try{if(data.get('error').asText())throw Error('DB_ERROR');const h=Sql.planHybrid(p,d,bindings,{strict});out.hybrid_value=Sql.executeHybrid(h,()=>data.get('rows'),data.get('context')).dump();}catch(e){out.hybrid_error=e.code??String(e);}}n++;
  console.log(JSON.stringify(out));
 }
}

File: tools/adversarial/translate.py
"""Standalone adapter that also exercises the installed Python wheel."""
import json
import os
import io
from sel import compile
from sel.sql import Binding, Sql

b={t.upper():Binding.relation(t,t,{c.upper():Binding.column(c,t,'TEXT' if c=='cat' else 'NUM') for c in ('id','cat','v','fk')}) for t in ('r','s')}
if os.getenv('AUDIT_UNIQUE_KEY'): b['R'] = b['R'].with_unique_key('id')
replay=io.StringIO(os.environ['AUDIT_REPLAY_DATA']) if 'AUDIT_REPLAY_DATA' in os.environ else open('tools/adversarial/replay-python-wheel.sel') if os.getenv('AUDIT_REPLAY') else None
with (io.StringIO(os.environ['AUDIT_QUERY']) if 'AUDIT_QUERY' in os.environ else open('tools/adversarial/queries.sel')) as lines:
 for i,source in enumerate(lines):
  for d in ('sqlite','postgresql','mariadb'):
   for strict in (False,True):
    out=dict(i=i,d=d,strict=strict);p=compile(source)
    try:
     f=Sql.translate_statement(p,d,b,{'strict':strict})
     out.update(sql=f.as_statement(),params_sql=f.as_statement('params'),params=[v.as_text() for v in f.bindings()],caveats=f.caveats)
    except Exception as e:out['error']=getattr(e,'code',str(e))
    try:
     h=Sql.plan_hybrid(p,d,b,{'strict':strict})
     out.update(plan='pure_sql' if h.pure_sql else 'pure_memory' if h.pure_memory else 'hybrid',prefix=h.sql_statement.as_statement() if h.sql_statement else None)
     if h.sql_statement: out.update(prefix_params_sql=h.sql_statement.as_statement('params'), prefix_params=[v.as_text() for v in h.sql_statement.bindings()])
    except Exception as e:out['plan_error']=getattr(e,'code',str(e))
    if replay:
     data=compile(next(replay)).run()
     try:
      if data.get('error').as_text():raise RuntimeError('DB_ERROR')
      h=Sql.plan_hybrid(p,d,b,{'strict':strict})
      out['hybrid_value']=Sql.execute_hybrid(h,lambda *_:data.get('rows'),data.get('context')).dump()
     except Exception as e:out['hybrid_error']=getattr(e,'code',str(e))
    print(json.dumps(out))
